libgit/repo.go

```diff
@@ -0,0 +1,153 @@
+type InvalidRepoNameError struct {
+	name string
+}
+
+func (e InvalidRepoNameError) Error() string {
+	return fmt.Sprintf("Invalid repo name %q", e.name)
+}
+
+func createNewRepoAndID(
+	ctx context.Context, config libkbfs.Config, tlfHandle *libkbfs.TlfHandle,
+	repoName string, fs *libfs.FS) (ID, error) {
+	if !checkValidRepoName(repoName, config) {
+		return NullID, errors.WithStack(InvalidRepoNameError{repoName})
+	}
+
+	// TODO: take a global repo lock here to make sure only one
+	// client generates the repo ID.
+	repoID, err := makeRandomID()
+	if err != nil {
+		return NullID, err
+	}
+	config.MakeLogger("").CDebugf(ctx,
+		"Creating a new repo %s in %s: repoID=%s",
+		repoName, tlfHandle.GetCanonicalPath(), repoID)
+	c := &Config{repoID, repoName}
+	buf, err := c.toBytes()
+	if err != nil {
+		return NullID, err
+	}
+	f, err := fs.Create(kbfsConfigName)
+	if err != nil {
+		return NullID, err
+	}
+	defer f.Close()
+	_, err = f.Write(buf)
+	if err != nil {
+		return NullID, err
+	}
+	return repoID, nil
+}
+
+// GetOrCreateRepoAndID returns a filesystem object rooted at the
+// specified repo, along with the stable repo ID.  If the repo hasn't
+// been created yet, it generates a new ID and creates the repo.  The
+// caller is responsible for syncing the FS and flushing the journal,
+// if desired.
+func GetOrCreateRepoAndID(
+	ctx context.Context, config libkbfs.Config, tlfHandle *libkbfs.TlfHandle,
+	repoName string, uniqID string) (*libfs.FS, ID, error) {
+	rootNode, _, err := config.KBFSOps().GetOrCreateRootNode(
+		ctx, tlfHandle, libkbfs.MasterBranch)
+	if err != nil {
+		return nil, NullID, err
+	}
+	normalizedRepoName := strings.ToLower(repoName)
+
+	lookupOrCreateDir := func(n libkbfs.Node, name string) (
+		libkbfs.Node, error) {
+		newNode, _, err := config.KBFSOps().Lookup(ctx, n, name)
+		switch errors.Cause(err).(type) {
+		case libkbfs.NoSuchNameError:
+			newNode, _, err = config.KBFSOps().CreateDir(ctx, n, name)
+			if err != nil {
+				return nil, err
+			}
+		case nil:
+		default:
+			return nil, err
+		}
+		return newNode, nil
+	}
+
+	repoDir, err := lookupOrCreateDir(rootNode, kbfsRepoDir)
+	if err != nil {
+		return nil, NullID, err
+	}
+	_, err = lookupOrCreateDir(repoDir, normalizedRepoName)
+	if err != nil {
+		return nil, NullID, err
+	}
+
+	fs, err := libfs.NewFS(
+		ctx, config, tlfHandle, path.Join(kbfsRepoDir, normalizedRepoName),
+		uniqID)
+	if err != nil {
+		return nil, NullID, err
+	}
+
+	f, err := fs.Open(kbfsConfigName)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, NullID, err
+	} else if os.IsNotExist(err) {
+		// Create a new repo ID.
+		repoID, err := createNewRepoAndID(ctx, config, tlfHandle, repoName, fs)
+		if err != nil {
+			return nil, NullID, err
+		}
+		return fs, repoID, nil
+	}
+	defer f.Close()
+
+	buf, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, NullID, err
+	}
+	c, err := configFromBytes(buf)
+	if err != nil {
+		return nil, NullID, err
```

Comment:
If we have a bad write for the config file, this means we won't be able to fix it without asking user to cd into the `.git` dir. But it seems simply overwriting it is a bit risky too. What do you think if we rename the old file and create a new one in this case?

Reply:
If we had an error while writing, then we'll never flush the journal (since we're in single-op mode).  So I don't see how a bad config file could happen.  I guess a bug in some older version of KBFS could have led to it.  But it's not as simple as recreating a new file, because we have to have the same stable ID and Name that we had before, and we have no other way of getting that info back automatically.  In the unlikely event of a problem, it seems to me like a manual fix would be best.

Reply:
Sounds good; thanks for the explanation! Good point on journal flush only happening with good writes. I was thinking about corrupted writes when config is created, but I guess there's no good way to know if it's first write either.